### PR TITLE
Fix receipt not found on block reorg

### LINKED ISSUE #4179: Missing Receipts on Ropsten
**Describe the bug**
When calling `eth_getTransactionReceipt` against a recent ropsten block I get a `null` response however, txHashes around it work fine, also asking for the receipts for the entire block also works and returns the hash in question.

```
{"jsonrpc": "2.0", "id": 1, "method": "eth_getTransactionReceipt", "params": ["0x5254301be6227507c6e5ae34b70ec0bcbc867f376cc4b24690f6fbc04882b742"]}
```

`{"jsonrpc": "2.0", "result": null, "id": 1}`

vs.

```
{"jsonrpc": "2.0", "id": 1, "method": "parity_getBlockReceipts", "params": [12422858]}
```

```
{"jsonrpc": "2.0", "result": [{"transactionHash": "0x5254301be6227507c6e5ae34b70ec0bcbc867f376cc4b24690f6fbc04882b742", "transactionIndex":    
"0x0", "blockHash": "0x2938464d37b08edcb3c4611e03c7a79b27cb907e9260d37a7c8cc487e3bebcc6", "blockNumber": "0xbd8eca", "cumulativeGasUsed":      
"0x1ac7c", "gasUsed": "0x1ac7c", "effectiveGasPrice": "0xb2d05e07", "from": "0xb0b95fd186e7d5266fa7459351b3eb75bc362409", "to":                
"0x4826e4ddeee65bc27034602a6050486fbe920707", "contractAddress": null, "logs": [{"removed": false, "logIndex": "0x0", "transactionIndex":      
"0x0", "transactionHash": "0x5254301be6227507c6e5ae34b70ec0bcbc867f376cc4b24690f6fbc04882b742", "blockHash":                                   
"0x2938464d37b08edcb3c4611e03c7a79b27cb907e9260d37a7c8cc487e3bebcc6", "blockNumber": "0xbd8eca", "address":                                    
"0x3f9f36ecb03ded97328fc7747292d6d03dfe3d46", "data": "0x00000000000000000000000000000000000000000000040ae94ee665e4e046c00000000000000000000000
004826e4ddeee65bc27034602a6050486fbe92070700000000000000000000000000000000000000000000000000000000003af000",
redacted
```

As you can see `parity_getBlockReceipts` knows all about `0x5254301be6227507c6e5ae34b70ec0bcbc867f376cc4b24690f6fbc04882b742` and returns stuff, but for whatever reason `eth_getTransactionReceipt` is like 🤷 

I tried upgrading to the latest version as well. Didn't change the output.

I also tried running with `"FixReceipts": true` in the config but no change.

### PR DESCRIPTION
**Pull request by asdacap**
Fixes #4179

Fix receipt not found issue when the tx hash pointed to a non-main block. Reorg can cause this issue, confirmed with a custom hive test.

https://github.com/ethereum/hive/pull/580/commits/d4108f55303ea3ecf647cef188a58bb93e4bfbb2

## Changes:
- On `BlockAddedToMain` event, call `EnsureCanonical` which will re-set the tx hash to block hash mapping to the canonical block.
- Also addded a repair routine on the receipt repair so that client don't have to re-sync.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing

- Confirmed test now pass. For reference, besu and geth was passing.
- Confirmed hive rpc module passed.
- Subscribed manually, confirmed log show up.

## Further comments (optional)

- The `removed` flag is not unset when switching to main block. I'd like to remove the field entirely, but it seems to be used on some gas calculation test and log subscription test. Not gonna mess arount that at the moment.

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs [hunk 1]
```diff
@@ -51,6 +53,7 @@ public PersistentReceiptStorage(IColumnsDb<ReceiptsColumns> receiptsDb, ISpecPro
             _receiptsRecovery = receiptsRecovery ?? throw new ArgumentNullException(nameof(receiptsRecovery));
             _blocksDb = _database.GetColumnDb(ReceiptsColumns.Blocks);
             _transactionDb = _database.GetColumnDb(ReceiptsColumns.Transactions);
+            _logger = logManager.GetClassLogger(GetType());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:56]: Is this logger ever used and needed?

**Comment by LukaszRozmej** [comment]
I am wondering if the main reason of the problem is that we are not going through main path of queue through blockprocessor, potentially yes. but still might be a problem after fork choice updated, that we need to keep track of. Will investigate more.

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:56]: Actually no. It was used before for debugging, but the logs was removed.

**Comment by asdacap** [comment]
It's more accurate to say that the tx mapping is not updated on reorg.

File: src/Nethermind/Nethermind.Init/Steps/RegisterRpcModules.cs [hunk 2]
```diff
@@ -203,12 +204,14 @@ public virtual async Task Execute(CancellationToken cancellationToken)
             WitnessRpcModule witnessRpcModule = new(_api.WitnessRepository, _api.BlockTree);
             rpcModuleProvider.RegisterSingle<IWitnessRpcModule>(witnessRpcModule);
             
+            ReceiptCanonicalityMonitor receiptCanonicalityMonitor = new(_api.BlockTree, _api.ReceiptStorage, _api.LogManager);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Init/Steps/RegisterRpcModules.cs:207]: What if RPC is disabled? This is too crucial to be here.

**Review by LukaszRozmej — APPROVED** [review]
Moved initialization back to InitializeBlockchain. Also someone else might want to observe that event.
Other than that added some syntactic sugar cosmetic refactors.
I would like to @marcindsobczak to take a look before merge, but looks good to me.

File: src/Nethermind/Nethermind.Init/Steps/RegisterRpcModules.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Init/Steps/RegisterRpcModules.cs]: I moved it back

**Review by marcindsobczak — APPROVED** [review]
Code looks fine, I tested log subscription manually and didn't find anything suspicious as well

**Pull request merged**